### PR TITLE
Add poll-endpoint for push token

### LINKED ISSUE #2262: Allow polling for push token challenges
We need to add an additional token specific endpoint, so that the push token can also poll for challenges.

See

https://github.com/privacyidea/privacyidea/wiki/concept:-pushtoken-poll

# Todo

* [x] Specifiy parameters for the endpoint /ttype/push
* [x] Implementation
* [ ] documentation
* [x] tests.

## Parameters

The endpoint is already used for
* 2nd enrollment step
* sending response for a challenge to privacyidea

When the smartphone answers a challenge it already sends: **serial, nonce and signature**.

So in this case, when polling, we use the parameters **serial, timestamp,** ~~firebase_token~~, **signature**

This is usually sent via the GET method.

The server does the following:

1. look at the timestamp. Fail, if timestamp is off by (configurable) one minute
2. (Currently we would not check for replay attacks, since then we would have to store the last timestamp)
3. Get public key of smartphone by serial from database table ``token``
4. check signature of serial and timestamp
5. Get valid challenges from the challenge table in the database by serial (only not answered challenges) and send this data https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/tokens/pushtoken.py#L541 plus signature as JSON response

## Response

The response is a standard privacyIDEA response with a list of challenge data, as it would be sent via firebase.

It can be a list, since in theory the user could try to login at several web applications at the same time, which would create several challenges "Do you want to login at ownCloud", "Do you want to login at simpleSAML"...

~~~~json
....
"result" : {"status": True,
                "value": [ {"nonce": challenge,
                                "question": message_on_mobile,
                                "serial": self.token.serial,
                                "title": title,
                                "sslverify": sslverify,
                                "url": url,
                                "signature": <signature over signature string> },
                                { next challenge pack }
                              ]}
....
~~~~

This is similar to what happens in the Push Tokenclass method ``create_challenge``.
The only difference is, that the data is not created but read from the challenge table.

## Possible errors

* No challenge available: Status: True, Value: empty list
* Timestamp out of sync: Status: False, Error message: Timestamp out of sync
* Serial does not exist: Status: False, Error message: Signature does not match serial number
* Signature failure: Status: False, Error message: Signature does not match serial number

### PR DESCRIPTION
**Pull request by plettich**
The endpoint uses the ttype API-endpoint to poll for current challenges.

Fixes #2262

### CODE REVIEW

File: privacyidea/lib/tokens/pushtoken.py [hunk 1]
```diff
@@ -144,34 +157,77 @@ def create_push_token_url(url=None, ttl=10, issuer="privacyIDEA", serial="mylabe
...
-    The PUSH token uses the firebase service to send challenges to the
-    users smartphone. The user confirms on the smartphone, signes the
+    The :ref:`push_token` uses the firebase service to send challenges to the
+    users smartphone. The user confirms on the smartphone, signs the
     challenge and sends it back to privacyIDEA.
 
     The enrollment occurs in two enrollment steps:
 
-    # Step 1
+    Step 1:
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py:213]: I don't know where You looked at it but currently this is completely un-formated: https://privacyidea.readthedocs.io/en/latest/modules/lib/tokentypes/push.html

File: privacyidea/lib/tokens/pushtoken.py [hunk 2]
```diff
@@ -66,16 +73,22 @@
...
 class PUSH_ACTION(object):
     FIREBASE_CONFIG = "push_firebase_configuration"
     MOBILE_TEXT = "push_text_on_mobile"
     MOBILE_TITLE = "push_title_on_mobile"
     SSL_VERIFY = "push_ssl_verify"
     WAIT = "push_wait"
+    ALLOW_POLLING = "push_allow_polling"
+
+
+PushAllowPolling = Enum('PushAllowPolling', 'allow deny token')
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py:91]: Yeah, i like enums, they have some nice properties but i'll revert back to constant classes

File: privacyidea/lib/tokens/pushtoken.py [hunk 3]
```diff
@@ -394,84 +461,187 @@ def get_init_detail(self, params=None, user=None):
...
+            else:
+                now = datetime.utcnow()
+            if not (now - td <= ts <= now + td):
+                raise privacyIDEAError('Timestamp {0!s} not in valid range.'.format(timestamp))
+            # now check the signature
+            # first get the token
+            try:
+                tok = get_one_token(serial=serial, tokentype='push')
+                # If the POLLING_ALLOWED tokeninfo is not set, we allow polling per default
+                if not is_true(tok.get_tokeninfo(POLLING_ALLOWED, default='True')):
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py:595]: You are right, forgot to check whether the Policy is set to 'token' before checking the token...

File: doc/policies/authentication.rst [hunk 4]
```diff
@@ -419,6 +419,38 @@ Sensible numbers might be 10 or 20 seconds.
...
+.. _policy_auth_push_allow_poll:
+
+push_allow_polling
+~~~~~~~~~~~~~~~~~~
+
+.. index:: push token
+
+type: string
+
+This policy configures if push token are allowed to poll the server for open
```
[the remark this answers is not in the document]
Comment (plettich) [reply doc/policies/authentication.rst]: Done in 14fcccce

File: privacyidea/api/lib/prepolicy.py [hunk 5]
```diff
@@ -1042,7 +1042,7 @@ def check_base_action(request=None, action=None, anonymous=False):
     :param request:
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/api/lib/prepolicy.py:1043]: Done in 14fcccce

File: privacyidea/lib/tokens/pushtoken.py [hunk 6]
```diff
@@ -144,34 +157,77 @@ def create_push_token_url(url=None, ttl=10, issuer="privacyIDEA", serial="mylabe
...
+    smartphone_data["signature"] = b32encode_and_unicode(signature)
+    return smartphone_data
+
+
 class PushTokenClass(TokenClass):
     """
-    The PUSH token uses the firebase service to send challenges to the
-    users smartphone. The user confirms on the smartphone, signes the
+    The :ref:`push_token` uses the firebase service to send challenges to the
+    users smartphone. The user confirms on the smartphone, signs the
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccce

File: privacyidea/lib/tokens/pushtoken.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: In 14fcccce i formatted it bold, the text for the steps is indented so they are pretty noticeable i think:
push-token-doc

File: privacyidea/lib/tokens/pushtoken.py [hunk 7]
```diff
@@ -144,34 +157,77 @@ def create_push_token_url(url=None, ttl=10, issuer="privacyIDEA", serial="mylabe
...
-    # Step 2
-
-    In the QR code is a URL, where the smartphone sends the remaining data for the enrollment.
-
-    POST https://yourprivacyideaserver/ttype/push
-     enrollment_credential=<some credential>
-     serial=<token serial>
-     fbtoken=<firebase token>
-     pubkey=<public key>
+          POST https://yourprivacyideaserver/ttype/push
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py [hunk 8]
```diff
@@ -394,84 +461,187 @@ def get_init_detail(self, params=None, user=None):
...
-        This endpoint is also used, if the smartphone sends the signed response
-        to the challenge during authentication
-        Parameters sent:
-            * serial
-            * nonce (which is the challenge)
-            * signature (which is the signed nonce)
+        - It is used for the 2nd enrollment step of the smartphone.
+          It accepts the following parameters::
 
+              POST https://yourprivacyideaserver/ttype/push
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py [hunk 9]
```diff
@@ -394,84 +461,187 @@ def get_init_detail(self, params=None, user=None):
...
+
+                  serial=<token serial>
+                  fbtoken=<firebase token>
+                  pubkey=<public key>
+
+        - It is also used when the smartphone sends the signed response
+          to the challenge during authentication. The following parameters ar accepted::
+
+              POST https://yourprivacyideaserver/ttype/push
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py [hunk 10]
```diff
@@ -394,84 +461,187 @@ def get_init_detail(self, params=None, user=None):
...
-
-            if challengeobject_list:
-                # There are valid challenges, so we check this signature
+                ts = isoparse(timestamp)
+            except (ValueError, TypeError) as _e:
+                log.debug('{0!s}'.format(traceback.format_exc()))
+                raise privacyIDEAError('Could not parse timestamp {0!s}. '
+                                       'ISO-Format required.'.format(timestamp))
+            # TODO: make time delta configurable
+            td = timedelta(minutes=1)
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py [hunk 11]
```diff
@@ -394,84 +461,187 @@ def get_init_detail(self, params=None, user=None):
...
+            if ts.tzinfo:
+                now = datetime.now(utc)
+            else:
+                now = datetime.utcnow()
+            if not (now - td <= ts <= now + td):
+                raise privacyIDEAError('Timestamp {0!s} not in valid range.'.format(timestamp))
+            # now check the signature
+            # first get the token
+            try:
+                tok = get_one_token(serial=serial, tokentype='push')
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py [hunk 12]
```diff
@@ -394,84 +461,187 @@ def get_init_detail(self, params=None, user=None):
...
+                # If the POLLING_ALLOWED tokeninfo is not set, we allow polling per default
+                if not is_true(tok.get_tokeninfo(POLLING_ALLOWED, default='True')):
+                    log.debug('Polling not allowed for pushtoken {0!s} due to '
+                              'tokeninfo.'.format(serial))
+                    raise PolicyError('Polling not allowed!')
+                pubkey_pem = tok.get_tokeninfo(PUBLIC_KEY_SMARTPHONE)
```
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 14fcccc

File: privacyidea/lib/tokens/pushtoken.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/pushtoken.py]: THanks a lot.

File: privacyidea/lib/tokens/pushtoken.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/pushtoken.py]: Thanks

File: privacyidea/lib/tokens/pushtoken.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/pushtoken.py]: OK, I think we had a misunderstanding. Please improve the docs this way, that it clarifies, if "token" is set, then all tokens are allowed, except where "POLLING_ALLOWED" evalutes to False.

File: privacyidea/lib/tokens/pushtoken.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/pushtoken.py]: thx

File: privacyidea/lib/tokens/pushtoken.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/pushtoken.py]: Cool. Again I mixed up rST and md.

File: privacyidea/lib/tokens/pushtoken.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/tokens/pushtoken.py]: Done in 516f073f

**Review by cornelinux — APPROVED** [review]

**Pull request merged**